cdap-common/src/main/java/co/cask/cdap/common/utils/Networks.java

```diff
@@ -94,4 +96,19 @@ public static String normalizeWebappDiscoveryName(String name) throws Unsupporte
 
     return URLEncoder.encode(name, Charsets.UTF_8.name());
   }
+
+
+  /**
+   * Get the filename part from a {@link URI}. For a URI "file:///tmp/foo.jar", "foo.jar" will be returned.
+   * For a URI "hdfs:///tmp/foo.jar#bar.jar", "bar.jar" will be returned.
+   *
+   * @param uri The {@link URI} to get filename from.
+   * @return filename.
+   */
+  public static String getURIFilename(URI uri) {
+    if ("hdfs".equals(uri.getScheme()) && uri.getFragment() != null) {
```

Comment:
This is wrong. Shouldn't check for `hdfs` specifically. Also, this method should simply return the filename part of the URI. It shouldn't be aware of the fragment, which only meaningful under the context of file localization (meaning only AbstractDistributedProgramRunner should be aware of that)